### PR TITLE
Improve follow/ufollow mutation

### PR DESCRIPTION
**Pull request by Tirokk**
> [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-16T11:00:37Z" title="Monday, September 16th 2019, 1:00:37 pm +02:00">Sep 16, 2019</time>_
_Merged <time datetime="2019-09-19T14:54:36Z" title="Thursday, September 19th 2019, 4:54:36 pm +02:00">Sep 19, 2019</time>_
---

## 🍰 Pullrequest
Update follow/unfollow mutation to return followed user

### Issues
- relates #1583 

### Todo
- [x] Update schema
- [x] Update spec
- [x] Update mutation usage in a client

### CODE REVIEW

**Comment by Tirokk** [comment]
> [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-16T11:01:57Z" title="Monday, September 16th 2019, 1:01:57 pm +02:00">Sep 16, 2019</time>_

---


FYI, the PR is going to be quite big to review at once. At the moment only the spec was refactored to use `apollo-server-testing`

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/follow.spec.js:1]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-16T19:34:00Z" title="Monday, September 16th 2019, 9:34:00 pm +02:00">Sep 16, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,10 +1,17 @@
-import { GraphQLClient } from 'graphql-request'
+import { createTestClient } from 'apollo-server-testing'
```


:heart_eyes_cat: you refactor our legacy backend tests!

@aonomike :point_up:

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/follow.spec.js:55]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-16T19:36:26Z" title="Monday, September 16th 2019, 9:36:26 pm +02:00">Sep 16, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -17,20 +24,42 @@ const mutationUnfollowUser = id => `
   }
 `
 
-beforeEach(async () => {
-  await factory.create('User', {
-    id: 'u1',
-    email: 'test@example.org',
-    password: '1234',
-  })
-  await factory.create('User', {
-    id: 'u2',
-    email: 'test2@example.org',
-    password: '1234',
+beforeAll(() => {
```


FYI @vbelolapotkov the `beforeAll` in most of our backend tests is *not only* a performance thing. There is also a bug in [graphql middlewares](https://github.com/prisma/graphql-middleware/issues/63#issuecomment-520786385) which will wrap another layer of middlewares around each resolver for every test case if you use `beforeEach`.

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/follow.spec.js:32]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-16T19:40:48Z" title="Monday, September 16th 2019, 9:40:48 pm +02:00">Sep 16, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -41,44 +70,57 @@ describe('follow', () => {
   describe('follow user', () => {
     describe('unauthenticated follow', () => {
       it('throws authorization error', async () => {
-        const client = new GraphQLClient(host)
-        await expect(client.request(mutationFollowUser('u2'))).rejects.toThrow('Not Authorised')
+        authenticatedUser = null
+        const { errors } = await mutate({
+          mutation: mutationFollowUser('u2'),
```


```suggestion
          mutation: mutationFollowUser, variables
```
Make the `mutationFollowUser` a `const` instead of a method. The string interpolation originates from a time when we weren't familiar enough with GraphQL and we did not know how to use `variables`.

The pattern which I happen to love is to mutate a global `variables` object in a `beforeEach` of a scope. Then pass these `variables` into the call of `mutate` or `query`. In my opinion that's a good way to avoid false negatives.

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/follow.spec.js:80]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-16T19:42:26Z" title="Monday, September 16th 2019, 9:42:26 pm +02:00">Sep 16, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -41,44 +70,57 @@ describe('follow', () => {
   describe('follow user', () => {
     describe('unauthenticated follow', () => {
       it('throws authorization error', async () => {
-        const client = new GraphQLClient(host)
-        await expect(client.request(mutationFollowUser('u2'))).rejects.toThrow('Not Authorised')
+        authenticatedUser = null
+        const { errors } = await mutate({
+          mutation: mutationFollowUser('u2'),
+        })
+        expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
       })
     })
 
     it('I can follow another user', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u2'))
-      const expected = {
-        follow: true,
-      }
-      expect(res).toMatchObject(expected)
+      const { data: result } = await mutate({
+        mutation: mutationFollowUser(user2.id),
+      })
+
+      const expected = { follow: true }
```


```suggestion
      const expected = { name: 'Name of the followed user', followedBy: [ { name: 'Me' } ], followedByCurrentUser: true }
```

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/follow.spec.js:85]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-16T19:42:43Z" title="Monday, September 16th 2019, 9:42:43 pm +02:00">Sep 16, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -41,44 +70,57 @@ describe('follow', () => {
   describe('follow user', () => {
     describe('unauthenticated follow', () => {
       it('throws authorization error', async () => {
-        const client = new GraphQLClient(host)
-        await expect(client.request(mutationFollowUser('u2'))).rejects.toThrow('Not Authorised')
+        authenticatedUser = null
+        const { errors } = await mutate({
+          mutation: mutationFollowUser('u2'),
+        })
+        expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
       })
     })
 
     it('I can follow another user', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u2'))
-      const expected = {
-        follow: true,
-      }
-      expect(res).toMatchObject(expected)
+      const { data: result } = await mutate({
+        mutation: mutationFollowUser(user2.id),
+      })
+
+      const expected = { follow: true }
+      expect(result).toMatchObject(expected)
+
+      const {
+        data: { User },
+      } = await query({
```


Remove this whole workaround.

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/follow.spec.js:54]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-16T19:43:44Z" title="Monday, September 16th 2019, 9:43:44 pm +02:00">Sep 16, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -41,44 +70,57 @@ describe('follow', () => {
   describe('follow user', () => {
     describe('unauthenticated follow', () => {
       it('throws authorization error', async () => {
-        const client = new GraphQLClient(host)
-        await expect(client.request(mutationFollowUser('u2'))).rejects.toThrow('Not Authorised')
+        authenticatedUser = null
+        const { errors } = await mutate({
+          mutation: mutationFollowUser('u2'),
+        })
+        expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
       })
     })
 
     it('I can follow another user', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u2'))
-      const expected = {
-        follow: true,
-      }
-      expect(res).toMatchObject(expected)
+      const { data: result } = await mutate({
+        mutation: mutationFollowUser(user2.id),
+      })
+
+      const expected = { follow: true }
+      expect(result).toMatchObject(expected)
+
+      const {
+        data: { User },
+      } = await query({
+        query: `{
+          User(id: "${user2.id}") {
+            followedBy { id }
+            followedByCurrentUser
+          }
+        }`,
+      })
 
-      const { User } = await clientUser1.request(`{
-        User(id: "u2") {
-          followedBy { id }
-          followedByCurrentUser
-        }
-      }`)
       const expected2 = {
-        followedBy: [{ id: 'u1' }],
+        followedBy: [{ id: user1.id }],
         followedByCurrentUser: true,
       }
       expect(User[0]).toMatchObject(expected2)
     })
 
     it('I can`t follow myself', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u1'))
-      const expected = {
-        follow: false,
-      }
-      expect(res).toMatchObject(expected)
+      const { data: result } = await mutate({
+        mutation: mutationFollowUser(user1.id),
+      })
+      const expected = { follow: false }
```


```suggestion
      const expected = { follow: null }
```

I think it's good practice to respond with `null` when nothing has been mutated in any way (e.g. nothing has been deleted).

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/follow.spec.js:133]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-16T19:44:11Z" title="Monday, September 16th 2019, 9:44:11 pm +02:00">Sep 16, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -41,44 +70,57 @@ describe('follow', () => {
   describe('follow user', () => {
     describe('unauthenticated follow', () => {
       it('throws authorization error', async () => {
-        const client = new GraphQLClient(host)
-        await expect(client.request(mutationFollowUser('u2'))).rejects.toThrow('Not Authorised')
+        authenticatedUser = null
+        const { errors } = await mutate({
+          mutation: mutationFollowUser('u2'),
+        })
+        expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
       })
     })
 
     it('I can follow another user', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u2'))
-      const expected = {
-        follow: true,
-      }
-      expect(res).toMatchObject(expected)
+      const { data: result } = await mutate({
+        mutation: mutationFollowUser(user2.id),
+      })
+
+      const expected = { follow: true }
+      expect(result).toMatchObject(expected)
+
+      const {
+        data: { User },
+      } = await query({
+        query: `{
+          User(id: "${user2.id}") {
+            followedBy { id }
+            followedByCurrentUser
+          }
+        }`,
+      })
 
-      const { User } = await clientUser1.request(`{
-        User(id: "u2") {
-          followedBy { id }
-          followedByCurrentUser
-        }
-      }`)
       const expected2 = {
-        followedBy: [{ id: 'u1' }],
+        followedBy: [{ id: user1.id }],
         followedByCurrentUser: true,
       }
       expect(User[0]).toMatchObject(expected2)
     })
 
     it('I can`t follow myself', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u1'))
-      const expected = {
-        follow: false,
-      }
-      expect(res).toMatchObject(expected)
+      const { data: result } = await mutate({
+        mutation: mutationFollowUser(user1.id),
+      })
+      const expected = { follow: false }
+      expect(result).toMatchObject(expected)
 
-      const { User } = await clientUser1.request(`{
-        User(id: "u1") {
+      const {
+        data: { User },
+      } = await query({
+        query: `{
+        User(id: "${user1.id}") {
           followedBy { id }
           followedByCurrentUser
         }
-      }`)
+      }`,
+      })
```


Same, remove, as described :point_up:

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/follow.spec.js:69]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-16T19:45:27Z" title="Monday, September 16th 2019, 9:45:27 pm +02:00">Sep 16, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -87,32 +129,39 @@ describe('follow', () => {
     })
   })
   describe('unfollow user', () => {
+    beforeEach(async () => {
+      await mutate({ mutation: mutationFollowUser(user2.id) })
```


You can either setup the data through a mutation or you can do:

```js
await myUser.relateTo(anotherUser, 'following')
```

If `myUser` and `anotherUser` is a neode node (that's the return type of the factory method).

**Comment by Tirokk** [comment]
> [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-17T06:46:11Z" title="Tuesday, September 17th 2019, 8:46:11 am +02:00">Sep 17, 2019</time>_

---


@roschaefer thanks for review and suggestions. Will continue working on improvements.

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [reply backend/src/schema/resolvers/follow.spec.js:69]: > [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-17T06:55:22Z" title="Tuesday, September 17th 2019, 8:55:22 am +02:00">Sep 17, 2019</time>_

---


@roschaefer I wouldn't agree on this. Suggested improvement manipulates DB "directly" rather than through known server API, but we test API and we shouldn't reason about the DB or any other tools used by the server.

In the future relation name, `neode` model or some other properties can change which will break the test setup. So we'll have to get back and fix the test.

If we keep the current approach the test setup will break only if we change mutation definition. However, if this happens we'll have to rewrite all the test anyways.

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [reply backend/src/schema/resolvers/follow.spec.js:55]: > [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-17T07:01:21Z" title="Tuesday, September 17th 2019, 9:01:21 am +02:00">Sep 17, 2019</time>_

---


Good point, thanks for the info

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [reply backend/src/schema/resolvers/follow.spec.js:80]: > [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-17T07:07:02Z" title="Tuesday, September 17th 2019, 9:07:02 am +02:00">Sep 17, 2019</time>_

---


@roschaefer good suggestion. Definitely will use it with further steps :)

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [reply backend/src/schema/resolvers/follow.spec.js:54]: > [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-17T07:27:02Z" title="Tuesday, September 17th 2019, 9:27:02 am +02:00">Sep 17, 2019</time>_

---


@roschaefer probably idea is good however this change requires schema update because of the mutation result is defined as `Boolean!` value. As we're about to change the schema significantly with further steps this change doesn't bring a lot of value IMO at the moment.

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [reply backend/src/schema/resolvers/follow.spec.js:32]: > [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-17T08:31:30Z" title="Tuesday, September 17th 2019, 10:31:30 am +02:00">Sep 17, 2019</time>_

---


done

File: backend/src/schema/resolvers/follow.js
Comment (Tirokk) [inline backend/src/schema/resolvers/follow.js:31]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-18T22:03:54Z" title="Thursday, September 19th 2019, 12:03:54 am +02:00">Sep 19, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,51 +1,44 @@
+import { neode as getNeode } from '../../bootstrap/neo4j'
+
+const neode = getNeode()
+
 export default {
   Mutation: {
     follow: async (_object, params, context, _resolveInfo) => {
-      const { id, type } = params
-
-      const session = context.driver.session()
-      const transactionRes = await session.run(
-        `MATCH (node {id: $id}), (user:User {id: $userId})
-          WHERE $type IN labels(node) AND NOT $id = $userId
-          MERGE (user)-[relation:FOLLOWS]->(node)
-          RETURN COUNT(relation) > 0 as isFollowed`,
-        {
-          id,
-          type,
-          userId: context.user.id,
-        },
-      )
-
-      const [isFollowed] = transactionRes.records.map(record => {
-        return record.get('isFollowed')
-      })
+      const { id: followedId, type } = params
+      const { user: currentUser } = context
 
-      session.close()
+      if (type === 'User' && currentUser.id === followedId) {
+        return null
+      }
 
-      return isFollowed
+      const [user, followedNode] = await Promise.all([
+        neode.find('User', currentUser.id),
+        neode.find(type, followedId),
+      ])
+      await user.relateTo(followedNode, 'following')
+      return followedNode.toJson()
     },
 
     unfollow: async (_object, params, context, _resolveInfo) => {
-      const { id, type } = params
-      const session = context.driver.session()
+      const { id: followedId, type } = params
+      const { user: currentUser } = context
 
-      const transactionRes = await session.run(
-        `MATCH (user:User {id: $userId})-[relation:FOLLOWS]->(node {id: $id})
+      /*
+       * Note: Neode doesn't provide an easy method for retrieving or removing relationships.
+       * It's suggested to use query builder feature (https://github.com/adam-cowley/neode/issues/67)
+       * However, pure cypher query looks cleaner IMO
+       */
```


Agree!

Comment (Tirokk) [reply backend/src/schema/resolvers/follow.js:31]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-18T22:05:19Z" title="Thursday, September 19th 2019, 12:05:19 am +02:00">Sep 19, 2019</time>_

---


I haven't made a final decision about it, ie. if we should use it for production code or if we should only use it for our factories. :thinking: :woman_shrugging:

File: backend/src/schema/types/schema.gql
Comment (Tirokk) [inline backend/src/schema/types/schema.gql:35]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-18T22:06:20Z" title="Thursday, September 19th 2019, 12:06:20 am +02:00">Sep 19, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -32,9 +32,9 @@ type Mutation {
   # Unshout the given Type and ID
   unshout(id: ID!, type: ShoutTypeEnum): Boolean!
   # Follow the given Type and ID
-  follow(id: ID!, type: FollowTypeEnum): Boolean!
+  follow(id: ID!, type: FollowTypeEnum): User
```


:+1: nice, looks great!

If you need any help or another review, feel free to ask for it @vbelolapotkov.

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/follow.spec.js:153]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-18T22:08:13Z" title="Thursday, September 19th 2019, 12:08:13 am +02:00">Sep 19, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -40,84 +97,73 @@ afterEach(async () => {
 describe('follow', () => {
   describe('follow user', () => {
     describe('unauthenticated follow', () => {
-      it('throws authorization error', async () => {
-        const client = new GraphQLClient(host)
-        await expect(client.request(mutationFollowUser('u2'))).rejects.toThrow('Not Authorised')
+      test('throws authorization error', async () => {
+        authenticatedUser = null
+        const { errors, data } = await mutate({
+          mutation: mutationFollowUser,
+          variables,
+        })
+        expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
+        expect(data).toMatchObject({ follow: null })
       })
     })
 
-    it('I can follow another user', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u2'))
-      const expected = {
-        follow: true,
-      }
-      expect(res).toMatchObject(expected)
-
-      const { User } = await clientUser1.request(`{
-        User(id: "u2") {
-          followedBy { id }
-          followedByCurrentUser
-        }
-      }`)
-      const expected2 = {
-        followedBy: [{ id: 'u1' }],
+    test('I can follow another user', async () => {
+      const { data: result } = await mutate({
+        mutation: mutationFollowUser,
+        variables,
+      })
+      const expectedUser = {
+        name: user2.name,
+        followedBy: [{ id: user1.id, name: user1.name }],
         followedByCurrentUser: true,
       }
-      expect(User[0]).toMatchObject(expected2)
+      expect(result).toMatchObject({ follow: expectedUser })
     })
 
-    it('I can`t follow myself', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u1'))
-      const expected = {
-        follow: false,
-      }
-      expect(res).toMatchObject(expected)
-
-      const { User } = await clientUser1.request(`{
-        User(id: "u1") {
-          followedBy { id }
-          followedByCurrentUser
-        }
-      }`)
-      const expected2 = {
+    test('I can`t follow myself', async () => {
+      variables.id = user1.id
+      const { data: result } = await mutate({ mutation: mutationFollowUser, variables })
+      const expectedResult = { follow: null }
+      expect(result).toMatchObject(expectedResult)
+
+      const { data } = await query({
+        query: userQuery,
+        variables: { id: user1.id },
+      })
+      const expectedUser = {
         followedBy: [],
         followedByCurrentUser: false,
       }
-      expect(User[0]).toMatchObject(expected2)
+      expect(data).toMatchObject({ User: [expectedUser] })
     })
   })
   describe('unfollow user', () => {
+    beforeEach(async () => {
+      variables = {
+        id: user2.id,
+        type: 'User',
+      }
+      await mutate({ mutation: mutationFollowUser, variables })
+    })
+
     describe('unauthenticated follow', () => {
-      it('throws authorization error', async () => {
-        // follow
-        await clientUser1.request(mutationFollowUser('u2'))
-        // unfollow
-        const client = new GraphQLClient(host)
-        await expect(client.request(mutationUnfollowUser('u2'))).rejects.toThrow('Not Authorised')
+      test('throws authorization error', async () => {
+        authenticatedUser = null
+        const { errors, data } = await mutate({ mutation: mutationUnfollowUser, variables })
```


You could also do this:
```js
await expect(mutate({ mutation: mutationUnfollowUser, variables })).resolves.toMatchObject({
  errors: [{'message': 'Not Authorised!'}],
  data: { unfollow: null },
})

Comment (Tirokk) [reply backend/src/schema/resolvers/follow.spec.js:153]: > [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-19T07:16:15Z" title="Thursday, September 19th 2019, 9:16:15 am +02:00">Sep 19, 2019</time>_

---


Thanks, I'm aware of it but I personally prefer separating a call for the method under test and assertions for the result. However, I'm fine to change it if there is a pattern in the codebase, 
@roschaefer please, let me know if this is a case.

File: backend/src/schema/types/schema.gql
Comment (Tirokk) [reply backend/src/schema/types/schema.gql:35]: > [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-19T07:21:07Z" title="Thursday, September 19th 2019, 9:21:07 am +02:00">Sep 19, 2019</time>_

---


@roschaefer actually this looks a little bit weird, especially `FollowTypeEnum` argument. I suppose this was done to reuse the mutation in the future to allow following other entities like posts or something else.

In that case, I'd expect some new `Followable` type (e.g. union type) or interface is returned. What do you think?

**Comment by Tirokk** [comment]
> [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-19T09:12:37Z" title="Thursday, September 19th 2019, 11:12:37 am +02:00">Sep 19, 2019</time>_

---


@roschaefer this one is ready for review

File: backend/src/schema/resolvers/follow.spec.js
Comment (Tirokk) [reply backend/src/schema/resolvers/follow.spec.js:153]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-19T14:26:56Z" title="Thursday, September 19th 2019, 4:26:56 pm +02:00">Sep 19, 2019</time>_

---


It's the pattern I stick to right now, but totally up to you! I thought you would be interested to see this.

Comment (Tirokk) [reply backend/src/schema/resolvers/follow.spec.js:153]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-19T14:27:37Z" title="Thursday, September 19th 2019, 4:27:37 pm +02:00">Sep 19, 2019</time>_

---


I just fond `toMatchObject` quite robust for testing a graphql response with very helpful error messages.

File: webapp/pages/profile/_id/_slug.vue
Comment (Tirokk) [inline webapp/pages/profile/_id/_slug.vue:426]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-19T14:52:49Z" title="Thursday, September 19th 2019, 4:52:49 pm +02:00">Sep 19, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -402,6 +403,27 @@ export default {
       this.resetPostList()
       this.$apollo.queries.Post.refetch()
     },
+    optimisticFollow({ followedByCurrentUser }) {
+      /*
+       * Note: followedByCountStartValue is updated to avoid counting from 0 when follow/unfollow
+       */
+      this.followedByCountStartValue = this.user.followedByCount
+      const currentUser = this.$store.getters['auth/user']
+      if (followedByCurrentUser) {
+        this.user.followedByCount++
+        this.user.followedBy = [currentUser, ...this.user.followedBy]
+      } else {
+        this.user.followedByCount--
+        this.user.followedBy = this.user.followedBy.filter(user => user.id !== currentUser.id)
+      }
+      this.user.followedByCurrentUser = followedByCurrentUser
+    },
+    updateFollow({ followedByCurrentUser, followedBy, followedByCount }) {
+      this.followedByCountStartValue = this.user.followedByCount
+      this.user.followedByCount = followedByCount
+      this.user.followedByCurrentUser = followedByCurrentUser
+      this.user.followedBy = followedBy
+    },
```


Wow, this is really sophisticated. I would have been fine with updating the user the pessimistic way™. This is really advanced and snappy.

File: backend/src/schema/types/schema.gql
Comment (Tirokk) [reply backend/src/schema/types/schema.gql:35]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-19T18:04:48Z" title="Thursday, September 19th 2019, 8:04:48 pm +02:00">Sep 19, 2019</time>_

---


This makes no sense at all. I guess the `FollowTypeEnum` was supposed to include `Organization`s as well. Something that we decided to not implement any soon. @vbelolapotkov do you want to land another quick contribution and remove `FollowTypeEnum`?

Comment (Tirokk) [reply backend/src/schema/types/schema.gql:35]: > [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-19T19:41:49Z" title="Thursday, September 19th 2019, 9:41:49 pm +02:00">Sep 19, 2019</time>_

---


@roschaefer will be happy to help here as well. I'd rename the mutation as well to `(un)followUser`

Comment (Tirokk) [reply backend/src/schema/types/schema.gql:35]: > [vbelolapotkov](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-20T14:01:42Z" title="Friday, September 20th 2019, 4:01:42 pm +02:00">Sep 20, 2019</time>_

---


@roschaefer here it is #1660 , please review when you have time

**Review by Mogge — APPROVED** [review]
Perfect!

Perfect

**Pull request closed without merging**
